src/bitcoind/interface.rs

```diff
@@ -89,64 +87,57 @@ impl BitcoinD {
             node_client,
             watchonly_client,
             cpfp_client,
+            retries: 0,
         })
     }
 
-    // Reasonably try to be robust to possible spurious communication error.
-    fn handle_error(&self, e: jsonrpc::Error, start: Instant) -> Result<(), BitcoindError> {
-        let now = Instant::now();
-
-        match e {
-            jsonrpc::Error::Transport(ref err) => {
-                log::error!("Transport error when talking to bitcoind: '{}'", err);
-
-                // This is *always* a simple_http::Error. Rule out the error that can
-                // not occur after startup (ie if we encounter them it must be startup
-                // and we better be failing quickly).
-                let any_err = err as &dyn Any;
-                if let Some(http_err) = any_err.downcast_ref::<HttpError>() {
-                    match http_err {
-                        HttpError::InvalidUrl { .. } => return Err(BitcoindError::Server(e)),
-                        // FIXME: allow it to be unreachable for a handful of seconds,
-                        // but not at startup!
-                        HttpError::SocketError(_) => return Err(BitcoindError::Server(e)),
-                        HttpError::HttpParseError => {
-                            // Weird. Try again once, just in case.
-                            if now.duration_since(start) > Duration::from_secs(1) {
-                                return Err(BitcoindError::Server(e));
+    /// Set the retry limit (number of times we'll retry a request to bitcoind upon specific failures).
+    pub fn with_retry_limit(mut self, retry_limit: usize) -> Self {
+        self.retries = retry_limit;
+        self
+    }
+
+    /// Wrapper to retry a request sent to bitcoind upon IO failure
+    /// according to the configured number of retries.
+    fn retry<T, R: Fn() -> Result<T, BitcoindError>>(
+        &self,
+        request: R,
+    ) -> Result<T, BitcoindError> {
+        let mut error: Option<BitcoindError> = None;
+        for i in 0..self.retries + 1 {
+            match request() {
+                Ok(res) => return Ok(res),
+                Err(e) => match e {
+                    BitcoindError::Server(Error::Rpc(RpcError { code, .. })) => {
+                        // Retry in case of Warming up error
+                        // https://github.com/bitcoin/bitcoin/blob/dca80ffb45fcc8e6eedb6dc481d500dedab4248b/src/rpc/protocol.h#L49
+                        if code == -28 {
+                            error = Some(e);
+                        } else {
+                            return Err(e);
+                        }
+                    }
```

Comment:
nit: we have a helper which does exactly this already

https://github.com/revault/revaultd/blob/457ff85dc40834ccf9e78ced25547a77d7278ad5/src/bitcoind/mod.rs#L38-L47

Reply:
It's fine as is. But i was thinking (untested):
```rust
BitcoindError::Server(_) => {
    if e.is_warming_up() {
        ...
    } else {
        ...
    }
}
```